(GH-1845) Add `bolt-defaults.yaml` configuration file (#1893)
puppetlabs/bolt: documentation/configuring_bolt.md

# Configuring Bolt

You can configure Bolt's options and features at a project level, a user level,
or a system-wide level. 
or a system-wide level.

> Some options and features you can configure include:

Reviewer: I really like the layout of this file - it's clear, well structured, and easy to follow

All configurable options in Bolt can be set at the project level. Configuration
that should only apply to the current project should be set at this level. Most
of the time, you will only need to set configuration at the project level.

Project-level configuration can be set in three locations. Bolt configuration
can be set in a `bolt-project.yaml` file, inventory configuration can be set in
an `inventory.yaml` file, and all configuration can be set in a `bolt.yaml`
file.

> **Note:** Project configuration files are loaded from the [Bolt project
> directory](bolt_project_directories.md). The default project directory is
> `~/.puppetlabs/bolt/`.

The preferred method for setting project-level configuration is to use a
combination of `bolt-project.yaml` and `inventory.yaml` files. This maintains

> a clear distinction between Bolt configuration and inventory configuration.

Reviewer: Note about deprecation?

Reviewer: I like the note below. I guess my only concern right here is that people will read this far, stop reading, create a bolt.yaml because it seems easier (1 file instead of 2), then have it get dropped in a few months.

### `bolt-project.yaml`

> **Note:** The `bolt-project.yaml` file is experimental and is subject to
> change. You can read more about Bolt projects in [Experimental
> features](experimental_features.md).

**Filepath:** `<project-directory>/bolt-project.yaml`

The project configuration file supports options that configure how Bolt
behaves, such as how many threads it can use when running commands on targets,
and configures different components of the project, such as a list of plans
and tasks that are visible to the user. Any directory containing a

> `bolt-project.yaml` file is automatically considered a [Project

Reviewer: I'm not sure how to convey this is only relevant if you're executing in that directory....maybe that's already obvious?

directory](bolt_project_directories.md).

Project configuration files take precedence over `bolt.yaml` files. If a
project directory contains both files, Bolt will only load and read
configuration from `bolt-project.yaml`.

The `bolt-project.yaml` file supports most of the options that a `bolt.yaml`
file does, with the exception of the following inventory configuration options:

- `transport`
- `docker`
- `local`
- `pcp`
- `remote`
- `ssh`
- `winrm`

You can view a full list of the available options in [Bolt configuration
options](bolt_configuration_reference.md).

### `inventory.yaml`

**Filepath:** `<project-directory>/inventory.yaml`

The inventory file is a structured data file that contains groups of targets
that you can run Bolt commands on, as well as configuration for the transports
used to connect to the targets. Most projects will include an inventory file.

> Inventory configuration can be set at multiple levels in an inventory file

Reviewer: ```suggestion
Inventory configuration can be set on any group, including the implicit 'all' group, in an inventory file
```

Author: Add targets to this?

```
Inventory configuration can be set on any group, including the implicit 'all' group, and any target in an inventory file
```

Reviewer: I like it

under a `config` option. The following options can be set under this option:

- `transport`
- `docker`
- `local`
- `pcp`
- `remote`
- `ssh`
- `winrm`

You can read more about inventory files and the available options in
[Inventory files](inventory_file_v2.md).

### `bolt.yaml`

> **Note** The project-level `bolt.yaml` file is on the path towards
> deprecation and will be removed in a future version of Bolt. Use
> `bolt-project.yaml` and `inventory.yaml` files instead.

**Filepath:** `<project-directory>/bolt.yaml`

The Bolt configuration file can be used to set all available configuration
options, including default inventory configuration options. Any directory
containing a `bolt.yaml` file is automatically considered a [Project
directory](bolt_project_directories.md).

You can view a full list of the available options in [Bolt configuration
options](bolt_configuration_reference.md).

## User-level configuration

> Most configurable options in Bolt can be set at the user level. Configuration

Reviewer: ```suggestion
Configuration
```

Reviewer: That line feels more confusing than illuminating, and raises questions like "What is user level configuration" and "Wait, but what options aren't included"

that should apply to all projects for a particular user should be set at this

> level. This might include paths to private keys, credentials for a plugin,

Reviewer: This and the next line are :100:

to lead to errors where content from another project is loaded.

### `bolt-defaults.yaml`

**Filepath:** `~/.puppetlabs/etc/bolt/bolt-defaults.yaml`

The defaults configuration file supports most of Bolt's configuration options,
with the exception of options that are project-specific such as `inventoryfile`
and `modulepath`.

The `bolt-defaults.yaml` file takes precedence over a `bolt.yaml` file in the
same directory. If the directory contains both files, Bolt will only load and 
read configuration from `bolt-defaults.yaml`.

You can view a full list of the available options in [`bolt-defaults.yaml`
options](bolt_defaults_reference.md).

### `bolt.yaml`

> **Note** The user-level `bolt.yaml` file is deprecated and will be removed
> in a future version of Bolt. Use a `bolt-defaults.yaml` file instead.

**Filepath:** `~/.puppetlabs/etc/bolt/bolt.yaml`

The Bolt configuration file can be used to set all available configuration
options, including project-specific configuration options.

You can view a full list of the available options in [Bolt configuration
options](bolt_configuration_reference.md).

## System-wide configuration

> Most configurable options in Bolt can be set at the system level. Configuration

Reviewer: ```suggestion
Configuration
```
